In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score


df = pd.read_csv(filepath_or_buffer="weatherAUS.csv")


In [2]:
df.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

In [ ]:
df.shapeEmbedded System Senior Engineer


(145460, 23)

In [4]:
column_names =  df.columns
column_names , f"Number of columns is ==> {len(column_names)}"

(Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
        'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
        'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
        'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
        'Temp3pm', 'RainToday', 'RainTomorrow'],
       dtype='object'),
 'Number of columns is ==> 23')

In [5]:
df.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

## Drop RISK_MM Variable

In [6]:
# it is already droped from this DB version 
# df.drop(['RISK_MM'], axis=1, inplace=1)

In [7]:
df.info()
# RainTomorrow ==> feature variable will be the output[y] that we need to pridict 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

## Types of Variables 
* Categorical variables of dtype=='O'==> object
* Numerical variables of dtype == 'f32,......'  

In [8]:
Categorical = [var for var in df.columns if df[var].dtype == 'O']

print(f"there are {len(Categorical)} Categorical variable")
print(f"the Categorical variables are {Categorical}")

there are 7 Categorical variable
the Categorical variables are ['Date', 'Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm', 'RainToday', 'RainTomorrow']


In [9]:
# view Categorical variables
df[Categorical].head()

Date Location WindGustDir WindDir9am WindDir3pm RainToday  \
0  2008-12-01   Albury           W          W        WNW        No   
1  2008-12-02   Albury         WNW        NNW        WSW        No   
2  2008-12-03   Albury         WSW          W        WSW        No   
3  2008-12-04   Albury          NE         SE          E        No   
4  2008-12-05   Albury           W        ENE         NW        No   

  RainTomorrow  
0           No  
1           No  
2           No  
3           No  
4           No

## Summary of categorical variables¶
* There is a date variable. It is denoted by Date column.
* There are 6 categorical variables. These are given by Location, WindGustDir, WindDir9am, WindDir3pm, RainToday and * RainTomorrow.
* There are two binary categorical variables - RainToday and RainTomorrow.
* RainTomorrow is the target variable.

## Explore problems within categorical variables¶
    First, I will explore the categorical variables.

    Missing values in categorical variables

In [10]:
# check missing values in categorical variables
# For each Categorical feature , we will count the missing values for it 
df[Categorical].isnull().sum()

Date                0
Location            0
WindGustDir     10326
WindDir9am      10566
WindDir3pm       4228
RainToday        3261
RainTomorrow     3267
dtype: int64

In [11]:
# printing Categorical variables only containing missing values

cat1 = [var for var in Categorical if df[var].isnull().sum() != 0]

print(f"{df[cat1].isnull().sum()}")


WindGustDir     10326
WindDir9am      10566
WindDir3pm       4228
RainToday        3261
RainTomorrow     3267
dtype: int64


We can see that there are only 4 categorical variables in the dataset which contains missing values. These are WindGustDir, WindDir9am, WindDir3pm and RainToday.

Frequency counts of categorical variables
Now, I will check the frequency counts of categorical variables.

In [12]:
# view frequency of categorical variables

for var in Categorical:
    print(df[var].value_counts())

Date
2017-06-24    49
2017-06-23    49
2017-06-22    49
2017-06-21    49
2017-06-20    49
              ..
2008-01-27     1
2008-01-28     1
2008-01-29     1
2008-01-30     1
2008-01-31     1
Name: count, Length: 3436, dtype: int64
Location
Canberra            3436
Sydney              3344
Adelaide            3193
Darwin              3193
Hobart              3193
Perth               3193
Melbourne           3193
Brisbane            3193
Albury              3040
MountGinini         3040
GoldCoast           3040
Wollongong          3040
MountGambier        3040
Launceston          3040
AliceSprings        3040
Albany              3040
Townsville          3040
Bendigo             3040
Cairns              3040
Ballarat            3040
Penrith             3039
Newcastle           3039
Tuggeranong         3039
PerthAirport        3009
Williamtown         3009
Cobar               3009
BadgerysCreek       3009
SydneyAirport       3009
Richmond            3009
Moree               3009
CoffsHarb

In [13]:
# view frequency distribution of categorical variables

for var in Categorical:
    print(f"{ df[var].value_counts() / np.float64(len(df)) }")

Date
2017-06-24    0.000337
2017-06-23    0.000337
2017-06-22    0.000337
2017-06-21    0.000337
2017-06-20    0.000337
                ...   
2008-01-27    0.000007
2008-01-28    0.000007
2008-01-29    0.000007
2008-01-30    0.000007
2008-01-31    0.000007
Name: count, Length: 3436, dtype: float64
Location
Canberra            0.023622
Sydney              0.022989
Adelaide            0.021951
Darwin              0.021951
Hobart              0.021951
Perth               0.021951
Melbourne           0.021951
Brisbane            0.021951
Albury              0.020899
MountGinini         0.020899
GoldCoast           0.020899
Wollongong          0.020899
MountGambier        0.020899
Launceston          0.020899
AliceSprings        0.020899
Albany              0.020899
Townsville          0.020899
Bendigo             0.020899
Cairns              0.020899
Ballarat            0.020899
Penrith             0.020892
Newcastle           0.020892
Tuggeranong         0.020892
PerthAirport        0.02

Number of labels: cardinality
The number of labels within a categorical variable is known as cardinality. A high number of labels within a variable is known as high cardinality. High cardinality may pose some serious problems in the machine learning model. So, I will check for high cardinality.

In [14]:
# check for cardinality in categorical variables

for var in Categorical:
    print(f"{var} contains {len(df[var].unique())}")

Date contains 3436
Location contains 49
WindGustDir contains 17
WindDir9am contains 17
WindDir3pm contains 17
RainToday contains 3
RainTomorrow contains 3


## Feature engineering of feature Date

In [15]:
df['Date'].dtypes

dtype('O')

we can see the data type of 'Date' variable is object, i will parse Data[coded as an object] into datetime foramt so i can seperate the Date into year,month,day 

In [16]:
# Parse Data variable ==> dates currently it is represented as string into datatime formate
df["Date"] = pd.to_datetime(df["Date"])


In [17]:
# Extract year from Date

# df["year"] = np.int64(df["Date"].dt.year)
df["year"] = df["Date"].dt.year
df['year'].head()

0    2008
1    2008
2    2008
3    2008
4    2008
Name: year, dtype: int32

In [18]:
# Extract Month from Date
df["month"] = df["Date"].dt.month
df["month"].head()

0    12
1    12
2    12
3    12
4    12
Name: month, dtype: int32

In [19]:
# Extract day from Date
df["day"] = df["Date"].dt.day
df["day"].head()

0    1
1    2
2    3
3    4
4    5
Name: day, dtype: int32

In [20]:
# we have added 3 columns to decompose the Data variable 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 26 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Date           145460 non-null  datetime64[ns]
 1   Location       145460 non-null  object        
 2   MinTemp        143975 non-null  float64       
 3   MaxTemp        144199 non-null  float64       
 4   Rainfall       142199 non-null  float64       
 5   Evaporation    82670 non-null   float64       
 6   Sunshine       75625 non-null   float64       
 7   WindGustDir    135134 non-null  object        
 8   WindGustSpeed  135197 non-null  float64       
 9   WindDir9am     134894 non-null  object        
 10  WindDir3pm     141232 non-null  object        
 11  WindSpeed9am   143693 non-null  float64       
 12  WindSpeed3pm   142398 non-null  float64       
 13  Humidity9am    142806 non-null  float64       
 14  Humidity3pm    140953 non-null  float64       
 15  

In [21]:
# we will remove Data variable since we replace it with year/month/day
df.drop(columns=["Date"], inplace=True)

In [22]:
# we will see the Date variable is no longer present anymore
df.head()

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0   Albury     13.4     22.9       0.6          NaN       NaN           W   
1   Albury      7.4     25.1       0.0          NaN       NaN         WNW   
2   Albury     12.9     25.7       0.0          NaN       NaN         WSW   
3   Albury      9.2     28.0       0.0          NaN       NaN          NE   
4   Albury     17.5     32.3       1.0          NaN       NaN           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Pressure3pm  Cloud9am  Cloud3pm  \
0           44.0          W        WNW  ...       1007.1       8.0       NaN   
1           44.0        NNW        WSW  ...       1007.8       NaN       NaN   
2           46.0          W        WSW  ...       1008.7       NaN       2.0   
3           24.0         SE          E  ...       1012.8       NaN       NaN   
4           41.0        ENE         NW  ...       1006.0       7.0       8.0   

   Temp9am  Temp3pm  RainToday  RainTomorrow  year  month  day  
0     16.9     21.8         No            No  2008     12    1  
1     17.2     24.3         No            No  2008     12    2  
2     21.0     23.2         No            No  2008     12    3  
3     18.1     26.5         No            No  2008     12    4  
4     17.8     29.7         No            No  2008     12    5  

[5 rows x 25 columns]

In [23]:
# now we can see the year/month/day columns
df.tail()

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
145455    Uluru      2.8     23.4       0.0          NaN       NaN   
145456    Uluru      3.6     25.3       0.0          NaN       NaN   
145457    Uluru      5.4     26.9       0.0          NaN       NaN   
145458    Uluru      7.8     27.0       0.0          NaN       NaN   
145459    Uluru     14.9      NaN       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am WindDir3pm  ...  Pressure3pm  \
145455           E           31.0         SE        ENE  ...       1020.3   
145456         NNW           22.0         SE          N  ...       1019.1   
145457           N           37.0         SE        WNW  ...       1016.8   
145458          SE           28.0        SSE          N  ...       1016.5   
145459         NaN            NaN        ESE        ESE  ...       1017.9   

        Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RainTomorrow  year  \
145455       NaN       NaN     10.1     22.4         No            No  2017   
145456       NaN       NaN     10.9     24.5         No            No  2017   
145457       NaN       NaN     12.5     26.1         No            No  2017   
145458       3.0       2.0     15.1     26.0         No            No  2017   
145459       8.0       8.0     15.0     20.9         No           NaN  2017   

        month  day  
145455      6   21  
145456      6   22  
145457      6   23  
145458      6   24  
145459      6   25  

[5 rows x 25 columns]

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 25 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Location       145460 non-null  object 
 1   MinTemp        143975 non-null  float64
 2   MaxTemp        144199 non-null  float64
 3   Rainfall       142199 non-null  float64
 4   Evaporation    82670 non-null   float64
 5   Sunshine       75625 non-null   float64
 6   WindGustDir    135134 non-null  object 
 7   WindGustSpeed  135197 non-null  float64
 8   WindDir9am     134894 non-null  object 
 9   WindDir3pm     141232 non-null  object 
 10  WindSpeed9am   143693 non-null  float64
 11  WindSpeed3pm   142398 non-null  float64
 12  Humidity9am    142806 non-null  float64
 13  Humidity3pm    140953 non-null  float64
 14  Pressure9am    130395 non-null  float64
 15  Pressure3pm    130432 non-null  float64
 16  Cloud9am       89572 non-null   float64
 17  Cloud3pm       86102 non-null

## Explore Location variable¶


In [25]:
# print number of labels in Location variable
print(f"Location contains {len(df.Location.unique())} lables")

Location contains 49 lables


In [26]:
# check labels in location variable

df.Location.unique()

array(['Albury', 'BadgerysCreek', 'Cobar', 'CoffsHarbour', 'Moree',
       'Newcastle', 'NorahHead', 'NorfolkIsland', 'Penrith', 'Richmond',
       'Sydney', 'SydneyAirport', 'WaggaWagga', 'Williamtown',
       'Wollongong', 'Canberra', 'Tuggeranong', 'MountGinini', 'Ballarat',
       'Bendigo', 'Sale', 'MelbourneAirport', 'Melbourne', 'Mildura',
       'Nhil', 'Portland', 'Watsonia', 'Dartmoor', 'Brisbane', 'Cairns',
       'GoldCoast', 'Townsville', 'Adelaide', 'MountGambier', 'Nuriootpa',
       'Woomera', 'Albany', 'Witchcliffe', 'PearceRAAF', 'PerthAirport',
       'Perth', 'SalmonGums', 'Walpole', 'Hobart', 'Launceston',
       'AliceSprings', 'Darwin', 'Katherine', 'Uluru'], dtype=object)

In [27]:
# check frequency distribution of values in Location variable
df.Location.value_counts()

Location
Canberra            3436
Sydney              3344
Adelaide            3193
Darwin              3193
Hobart              3193
Perth               3193
Melbourne           3193
Brisbane            3193
Albury              3040
MountGinini         3040
GoldCoast           3040
Wollongong          3040
MountGambier        3040
Launceston          3040
AliceSprings        3040
Albany              3040
Townsville          3040
Bendigo             3040
Cairns              3040
Ballarat            3040
Penrith             3039
Newcastle           3039
Tuggeranong         3039
PerthAirport        3009
Williamtown         3009
Cobar               3009
BadgerysCreek       3009
SydneyAirport       3009
Richmond            3009
Moree               3009
CoffsHarbour        3009
NorfolkIsland       3009
WaggaWagga          3009
Witchcliffe         3009
Watsonia            3009
Dartmoor            3009
Portland            3009
Sale                3009
MelbourneAirport    3009
Mildura         

## To convert catogerical variable into numerical one we will use one of the following:
* Lable Encoding
* One Hot Encoding
* Dummies 

In [28]:
# we have 49 labels but we will drop the first and keep 48[we will be able to deduce the 49th] and we need the model to faulty[just a way for optimization] 
# so we can overcome overfitting
# let's do One Hot Encoding of Location variable
# get k-1 dummy variables after One Hot Encoding
# preview the dataset with head() method


pd.get_dummies(df.Location, drop_first=True).head()

Albany  Albury  AliceSprings  BadgerysCreek  Ballarat  Bendigo  Brisbane  \
0   False    True         False          False     False    False     False   
1   False    True         False          False     False    False     False   
2   False    True         False          False     False    False     False   
3   False    True         False          False     False    False     False   
4   False    True         False          False     False    False     False   

   Cairns  Canberra  Cobar  ...  Townsville  Tuggeranong  Uluru  WaggaWagga  \
0   False     False  False  ...       False        False  False       False   
1   False     False  False  ...       False        False  False       False   
2   False     False  False  ...       False        False  False       False   
3   False     False  False  ...       False        False  False       False   
4   False     False  False  ...       False        False  False       False   

   Walpole  Watsonia  Williamtown  Witchcliffe  Wollongong  Woomera  
0    False     False        False        False       False    False  
1    False     False        False        False       False    False  
2    False     False        False        False       False    False  
3    False     False        False        False       False    False  
4    False     False        False        False       False    False  

[5 rows x 48 columns]

In [29]:
# re-viewing the Data set again
df.head()

Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0   Albury     13.4     22.9       0.6          NaN       NaN           W   
1   Albury      7.4     25.1       0.0          NaN       NaN         WNW   
2   Albury     12.9     25.7       0.0          NaN       NaN         WSW   
3   Albury      9.2     28.0       0.0          NaN       NaN          NE   
4   Albury     17.5     32.3       1.0          NaN       NaN           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Pressure3pm  Cloud9am  Cloud3pm  \
0           44.0          W        WNW  ...       1007.1       8.0       NaN   
1           44.0        NNW        WSW  ...       1007.8       NaN       NaN   
2           46.0          W        WSW  ...       1008.7       NaN       2.0   
3           24.0         SE          E  ...       1012.8       NaN       NaN   
4           41.0        ENE         NW  ...       1006.0       7.0       8.0   

   Temp9am  Temp3pm  RainToday  RainTomorrow  year  month  day  
0     16.9     21.8         No            No  2008     12    1  
1     17.2     24.3         No            No  2008     12    2  
2     21.0     23.2         No            No  2008     12    3  
3     18.1     26.5         No            No  2008     12    4  
4     17.8     29.7         No            No  2008     12    5  

[5 rows x 25 columns]

In [30]:
remaining_categorical = [var for var in df.columns if df[var].dtype == 'O' ]
print(f"{remaining_categorical}")


['Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm', 'RainToday', 'RainTomorrow']


## Explore WindGustDir variable¶


In [31]:
# print number of labels in WindGustDir variable
print(f"WindGustDir contains {len(df.WindGustDir.unique())} lables")

WindGustDir contains 17 lables


In [32]:
# check the lables in WindGustDir variable
df.WindGustDir.unique()

array(['W', 'WNW', 'WSW', 'NE', 'NNW', 'N', 'NNE', 'SW', nan, 'ENE',
       'SSE', 'S', 'NW', 'SE', 'ESE', 'E', 'SSW'], dtype=object)

In [33]:
# check frequency distribution of values in WindGustDir variable
df.WindGustDir.value_counts()

WindGustDir
W      9915
SE     9418
N      9313
SSE    9216
E      9181
S      9168
WSW    9069
SW     8967
SSW    8736
WNW    8252
NW     8122
ENE    8104
ESE    7372
NE     7133
NNW    6620
NNE    6548
Name: count, dtype: int64

In [34]:
# let's do One Hot Encoding of WindGustDir variable
# get k-1 dummy variables after One Hot Encoding
# preview the dataset with head() method


dummies = pd.get_dummies(df.WindGustDir, drop_first=True, dummy_na=True)
dummies.head()

ENE    ESE      N     NE    NNE    NNW     NW      S     SE    SSE  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False  False  False  False  False  False   
2  False  False  False  False  False  False  False  False  False  False   
3  False  False  False   True  False  False  False  False  False  False   
4  False  False  False  False  False  False  False  False  False  False   

     SSW     SW      W    WNW    WSW    NaN  
0  False  False   True  False  False  False  
1  False  False  False   True  False  False  
2  False  False  False  False   True  False  
3  False  False  False  False  False  False  
4  False  False   True  False  False  False

In [35]:
# CASE 1: axis=0 (Summing DOWN)
# This counts how many times 'N', 'S', 'W', or 'NaN' occurred. similiar to ==>[df.WindGustDir.value_counts()]
col_sums = dummies.sum(axis=0)
print("--- sum(axis=0) : Column Counts ---")
print(col_sums)

# CASE 2: axis=1 (Summing ACROSS)
# This sums the row. Since it's one-hot encoded,
# it confirms that each row has exactly one entry marked as 1.
row_sums = dummies.sum(axis=1)
print("\n--- sum(axis=1) : Row Totals ---")
print(row_sums)




--- sum(axis=0) : Column Counts ---
ENE     8104
ESE     7372
N       9313
NE      7133
NNE     6548
NNW     6620
NW      8122
S       9168
SE      9418
SSE     9216
SSW     8736
SW      8967
W       9915
WNW     8252
WSW     9069
NaN    10326
dtype: int64

--- sum(axis=1) : Row Totals ---
0         1
1         1
2         1
3         1
4         1
         ..
145455    0
145456    1
145457    1
145458    1
145459    1
Length: 145460, dtype: int64


we have NaN = 10326(missing values) needed to fixed 

In [36]:
print(f"{[var for var in df.columns if df[var].dtype == 'O']}")

['Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm', 'RainToday', 'RainTomorrow']


## Explore WindDir9am variable¶


In [37]:
# print number of labels in WindDir9am variable
print(f"WindGustDir contains {len(df.WindDir9am.unique())} lables")

WindGustDir contains 17 lables


In [38]:
# check the lables
df.WindDir9am.unique()

array(['W', 'NNW', 'SE', 'ENE', 'SW', 'SSE', 'S', 'NE', nan, 'SSW', 'N',
       'WSW', 'ESE', 'E', 'NW', 'WNW', 'NNE'], dtype=object)

In [39]:
# check the frequence distribution of the lables inside WindDir9am feature
df.WindDir9am.value_counts()

WindDir9am
N      11758
SE      9287
E       9176
SSE     9112
NW      8749
S       8659
W       8459
SW      8423
NNE     8129
NNW     7980
ENE     7836
NE      7671
ESE     7630
SSW     7587
WNW     7414
WSW     7024
Name: count, dtype: int64

In [ ]:
# let's do One Hot Encoding of WindDir9am variable
# get k-1 dummy variables after One Hot Encoding
# preview the dataset with head() method


WindDir9am_dummies =  pd.get_dummies(df.WindDir9am, drop_first=True, dummy_na=True).head()
WindDir9am_dummies


ENE    ESE      N     NE    NNE    NNW     NW      S     SE    SSE  \
0  False  False  False  False  False  False  False  False  False  False   
1  False  False  False  False  False   True  False  False  False  False   
2  False  False  False  False  False  False  False  False  False  False   
3  False  False  False  False  False  False  False  False   True  False   
4   True  False  False  False  False  False  False  False  False  False   

     SSW     SW      W    WNW    WSW    NaN  
0  False  False   True  False  False  False  
1  False  False  False  False  False  False  
2  False  False   True  False  False  False  
3  False  False  False  False  False  False  
4  False  False  False  False  False  False

In [45]:
#it is similair to .count_values()
WindDir9am_dummies.sum(axis=0) 

ENE    1
ESE    0
N      0
NE     0
NNE    0
NNW    1
NW     0
S      0
SE     1
SSE    0
SSW    0
SW     0
W      2
WNW    0
WSW    0
NaN    0
dtype: int64

## Explore RainToday variable¶
   this is the output lable  

In [48]:
# print number of labels in RainToday variable
print(f"WindGustDir contains {len(df.RainToday.unique())} lables")

WindGustDir contains 3 lables


In [49]:
# check the lables names
df.RainToday.unique()

array(['No', 'Yes', nan], dtype=object)

In [50]:
# check the distribuation for the lables inside the reponse feature
df.RainToday.value_counts()

RainToday
No     110319
Yes     31880
Name: count, dtype: int64

In [61]:
# RainToday_dummies = pd.get_dummies(df.RainToday, dummy_na=True)
# 	No	Yes	NaN
# 0	True	False	False
# 1	True	False	False
# 2	True	False	False
# 3	True	False	False
# 4	True	False	False
RainToday_dummies = pd.get_dummies(df.RainToday,drop_first=True, dummy_na=True)
RainToday_dummies.head()

Yes    NaN
0  False  False
1  False  False
2  False  False
3  False  False
4  False  False

In [62]:
RainToday_dummies.sum(axis=0)

Yes    31880
NaN     3261
dtype: int64

## Exploring the Numerical Values

In [64]:
numerical = [var for var in df.columns if df[var].dtype != 'O']

print(f"Number of numerical features is {len(numerical)}")
print(f"Numerical features is {numerical}")

Number of numerical features is 19
Numerical features is ['MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation', 'Sunshine', 'WindGustSpeed', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am', 'Temp3pm', 'year', 'month', 'day']


In [66]:
# view the numerical variables 
df[numerical].head()

MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
0     13.4     22.9       0.6          NaN       NaN           44.0   
1      7.4     25.1       0.0          NaN       NaN           44.0   
2     12.9     25.7       0.0          NaN       NaN           46.0   
3      9.2     28.0       0.0          NaN       NaN           24.0   
4     17.5     32.3       1.0          NaN       NaN           41.0   

   WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
0          20.0          24.0         71.0         22.0       1007.7   
1           4.0          22.0         44.0         25.0       1010.6   
2          19.0          26.0         38.0         30.0       1007.6   
3          11.0           9.0         45.0         16.0       1017.6   
4           7.0          20.0         82.0         33.0       1010.8   

   Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  year  month  day  
0       1007.1       8.0       NaN     16.9     21.8  2008     12    1  
1       1007.8       NaN       NaN     17.2     24.3  2008     12    2  
2       1008.7       NaN       2.0     21.0     23.2  2008     12    3  
3       1012.8       NaN       NaN     18.1     26.5  2008     12    4  
4       1006.0       7.0       8.0     17.8     29.7  2008     12    5

In [67]:
# Check the missing values in the numerical values
df[numerical].isnull().sum()

MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustSpeed    10263
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
year                 0
month                0
day                  0
dtype: int64

In [74]:
# Veiw summary statistics in numerical values

# df[numerical].describe()
#Round the precision number for better view [representation] 
round(df[numerical].describe(), 2)

MinTemp    MaxTemp   Rainfall  Evaporation  Sunshine  WindGustSpeed  \
count  143975.00  144199.00  142199.00     82670.00  75625.00      135197.00   
mean       12.19      23.22       2.36         5.47      7.61          40.04   
std         6.40       7.12       8.48         4.19      3.79          13.61   
min        -8.50      -4.80       0.00         0.00      0.00           6.00   
25%         7.60      17.90       0.00         2.60      4.80          31.00   
50%        12.00      22.60       0.00         4.80      8.40          39.00   
75%        16.90      28.20       0.80         7.40     10.60          48.00   
max        33.90      48.10     371.00       145.00     14.50         135.00   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
count     143693.00     142398.00    142806.00    140953.00    130395.00   
mean          14.04         18.66        68.88        51.54      1017.65   
std            8.92          8.81        19.03        20.80         7.11   
min            0.00          0.00         0.00         0.00       980.50   
25%            7.00         13.00        57.00        37.00      1012.90   
50%           13.00         19.00        70.00        52.00      1017.60   
75%           19.00         24.00        83.00        66.00      1022.40   
max          130.00         87.00       100.00       100.00      1041.00   

       Pressure3pm  Cloud9am  Cloud3pm    Temp9am    Temp3pm       year  \
count    130432.00  89572.00  86102.00  143693.00  141851.00  145460.00   
mean       1015.26      4.45      4.51      16.99      21.68    2012.77   
std           7.04      2.89      2.72       6.49       6.94       2.54   
min         977.10      0.00      0.00      -7.20      -5.40    2007.00   
25%        1010.40      1.00      2.00      12.30      16.60    2011.00   
50%        1015.20      5.00      5.00      16.70      21.10    2013.00   
75%        1020.00      7.00      7.00      21.60      26.40    2015.00   
max        1039.60      9.00      9.00      40.20      46.70    2017.00   

           month        day  
count  145460.00  145460.00  
mean        6.40      15.71  
std         3.43       8.79  
min         1.00       1.00  
25%         3.00       8.00  
50%         6.00      16.00  
75%         9.00      23.00  
max        12.00      31.00